### ISSUE TITLE
Using the RenderingServer to add a texture to the scene, the texture appears at bad position during one frame

### ISSUE THREAD

**Issue by kakatoto-collab** [godotengine/godot#94988]
### Tested versions

4.3-rc1 Mono Win64

### System information

Windows 11

### Issue description

As said in the title, I use the RenderingServer in order to add texture to the game scene. The problem is that during one frame, the texture appears at wrong position. More precisly, the wrong position is somewhere between the origin (0, 0) and the position it is supposed to be. After this wrong frame, the texture jumps to the expected position.

### Steps to reproduce

Launch the attached project and click in the game scene. You will see the green circle popping at bad position during one frame.

### Minimal reproduction project (MRP)

[BugSpriteJumpingAtInstanciation.zip](https://github.com/user-attachments/files/16446283/BugSpriteJumpingAtInstanciation.zip)

**Comment by clayjohn** [comment]
I don't have things set up to test with C#. But I strongly suspect the problem comes from when the sprite is added. You are creating the CanvasItem and issuing the draw command from the _process() function instead of doing the draw commands during the _draw() function. 

In the scene-level API Godot enforces that all drawing operations are done through the _draw() function. I suspect that there is an order of operations issue happening here. 

I tried recreating the MRP in GDScript, but I can't reproduce the issue:
```
extends Node2D

var sprite_sheet: Texture2D

# Called when the node enters the scene tree for the first time.
func _ready() -> void:
	sprite_sheet = load("res://icon.svg")


# Called every frame. 'delta' is the elapsed time since the previous frame.
func _process(delta: float) -> void:
	get_dev_input()


func get_dev_input():
	if Input.is_action_just_pressed("ui_click_left"):
		on_mouse_left_click()
	
func on_mouse_left_click():
	unsafe_test_draw()
	
func unsafe_test_draw():
	var src_rect = Rect2(Vector2(0, 0), Vector2(128, 128))
	var rid = RenderingServer.canvas_item_create()
	RenderingServer.canvas_item_set_parent(rid, get_canvas_item())

	var rect = Rect2(Vector2.ZERO, src_rect.size)
	sprite_sheet.draw_rect_region(rid, rect, src_rect)

	var transform_2d = Transform2D.IDENTITY
	transform_2d.origin = get_global_mouse_position()
	RenderingServer.canvas_item_set_transform(rid, transform_2d)
```

**Comment by kakatoto-collab** [comment]
Thanks for your reply, following your idea, I changed myc ode to this : 
'''
using Godot;
using System;

public partial class Game : Node2D
{
	public World world;
    public Texture2D Spritesheet;

    public override void _Ready()
	{
        Spritesheet = ResourceLoader.Load("res://Assets/Textures/Spritesheet.png") as Texture2D;

        world = GetNode("%World") as World;
	}

    public override void _Process(double delta)
	{
        QueueRedraw();
    }

    public override void _Draw()
    {
        GetDevInput();
    }

    private void GetDevInput()
    {
        if (Input.IsActionJustPressed("ui_click_left"))
        {
            OnMouseLeftClick();
        }

    }

    private void OnMouseLeftClick()
    {
        UnsafeTestDraw();
    }

    private void UnsafeTestDraw()
    {
        Rect2 srcRect = new Rect2(new Vector2(256, 32), new Vector2(16, 16));
        Rid rid = RenderingServer.CanvasItemCreate();
        RenderingServer.CanvasItemSetParent(rid, world.entityLayer.GetCanvasItem());
        Rect2 rect = new Rect2(Vector2.Zero, srcRect.Size);
        Spritesheet.DrawRectRegion(rid, rect, srcRect);

        Transform2D transform2D = Transform2D.Identity;
        transform2D.Origin = GetGlobalMousePosition();
        RenderingServer.CanvasItemSetTransform(rid, transform2D);
    }
}
'''

but the bug still occurs.

**Comment by kakatoto-collab** [comment]
![BugSpritesJUmpingOnInstanciation](https://github.com/user-attachments/assets/5011d745-f1f1-45cc-8edc-598e42925c23)

I createed a gif so that you can see what happens.

**Comment by clayjohn** [comment]
Hmmm, let's see if someone from the @godotengine/dotnet team can confirm if this is a C# issue

**Comment by raulsntos** [comment]
I'm able to reproduce with GDScript. I'll upload the ported MRP, but it's not too different from https://github.com/godotengine/godot/issues/94988#issuecomment-2261665421.

[BugSpriteJumpingAtInstanciationGD.zip](https://github.com/user-attachments/files/16537583/BugSpriteJumpingAtInstanciationGD.zip)

**Comment by clayjohn** [comment]
Thanks @raulsntos I can reproduce the issue with your MRP. I was able to narrow it down to physics interpolation. When you turn off physics interpolation, the issue goes away.

**Comment by clayjohn** [comment]
I took a quick look at how our CanvasItem code handles physics interpolation to avoid this issue. You have to flag for the RenderingServer that this item shouldn't be interpolated the frame it is created using: `RenderingServer.canvas_item_reset_physics_interpolation(rid)`

I confirmed in the MRP from RaulSntos that adding that function is all that is needed to fix the issue.

**Comment by kakatoto-collab** [comment]
I confirm that the bug is solved by deactivating Physics Interpolation in Project Settings.

**Comment by kakatoto-collab** [comment]
Sorry I closed by accident, I am not used to work on Github.

**Comment by cmagapu** [comment]
@kakatoto-collab I am new to Open Source and would like to contribute. Can I contribute to this?

**Comment by clayjohn** [comment]
@cmagapu yes! Please go ahead

**Comment by ghildim** [comment]
@kakatoto-collab Hey, I want to help with this project. How do I start?

**Comment by kakatoto-collab** [comment]
> @kakatoto-collab I am new to Open Source and would like to contribute. Can I contribute to this?

Yes sure, I am the best to help you to do so though.

**Comment by cmagapu** [comment]
@clayjohn so the task here is to document how the bug was solved (by deactivating Physics Interpolation in Project Settings), correct?
Does that mean I have to raise a PR here: https://github.com/godotengine/godot-docs ?

**Comment by AThousandShips** [comment]
No the documentation for classes is handled here

**Comment by ghildim** [comment]
> No the documentation for classes is handled here
So should add this to documentation or should we attempt to fix it?

**Comment by vas1l** [comment]
Hi, looking to make a contribution for OS project. How can I start?

**Comment by clayjohn** [comment]
> Hi, looking to make a contribution for OS project. How can I start?

A note needs to be added to the [documentation page](https://docs.godotengine.org/en/latest/tutorials/performance/using_servers.html) about servers letting users know that they should reset physics interpolation on the first frame an object is created using `RenderingServer.canvas_item_reset_physics_interpolation(rid)`

**Comment by Manik2607** [comment]
Can't we reset physics interpolation  internally inside the engine on the first frame an object is created using `RenderingServer.canvas_item_reset_physics_interpolation(rid)` as this behavior is abnormal?

**Comment by clayjohn** [comment]
> Can't we reset physics interpolation internally inside the engine on the first frame an object is created using `RenderingServer.canvas_item_reset_physics_interpolation(rid)` as this behavior is abnormal?

Maybe! I'm not sure how that would work, here is what the function currently does:
```
void RendererCanvasCull::canvas_item_reset_physics_interpolation(RID p_item) {
	Item *canvas_item = canvas_item_owner.get_or_null(p_item);
	ERR_FAIL_NULL(canvas_item);
	canvas_item->xform_prev = canvas_item->xform_curr;
}
```
You could look into whether its possible to detect when xform_curr is set initially and then init xform_prev at the same time. Or handle the first frame in some other way.  

The way its implemented now, it needs to be called after the position is set, but if you can figure out a way around that, it would be a nice improvement!

**Comment by furkanCalik7** [comment]
I want to contribute to Godot. Can I start here?

**Comment by mcdubhghlas** [comment]
Hello! I was trying to figure out this bug, but I don't think it is. It appears as expected behaviour.

I made this change in Game.gd:
```
func unsafe_test_draw():
	var src_rect: Rect2 = Rect2(Vector2(256, 32), Vector2(16, 16))
	var rid: RID = RenderingServer.canvas_item_create()
	RenderingServer.canvas_item_set_parent(rid, world.entity_layer.get_canvas_item())
	var rect: Rect2 = Rect2(Vector2(get_global_mouse_position()), src_rect.size)
	spritesheet.draw_rect_region(rid, rect, src_rect)
```

Here it is in a gif:
![bug](https://github.com/user-attachments/assets/61d8bfb6-3ecc-4350-8c4d-220769714ca9)

**Comment by jmii1236** [comment]
hi, i would like to contribute to solving this issue. how can i do that? this is for a uni project

**Comment by Aaron-Yng** [comment]
I agree with mcdubhghlas. I'm new to Godot and was not able to build the required version of Godot with dotnet in order to test the issue. However, from what I saw and what mcdubhghlas said it seems like the sprite appearing in between the origin and the specified coordinates is because physics interpolation causes the sprite to be drawn in between frames at an interpolated matrix (as said by Tuckertcs in the reddit thread below). In addition, the Godot docs also address this specific issue in the "Call reset_physics_interpolation() when teleporting objects" section. It states that to solve the issue stated in the post, you need to call reset_physics_interpolation() after setting the position/transform of the node and there won't be any unwanted streaking or intermediate sprites.

If this is right could this issue be closed? Thank you.

Reddit thread discussing physics interpolation:
https://www.reddit.com/r/godot/comments/1cxzhkp/are_there_any_downsides_of_physics_interpolation/

Godot docs regarding physics interpolation:
https://docs.godotengine.org/en/3.5/tutorials/physics/interpolation/using_physics_interpolation.html
